### ISSUE TITLE
Feature Request - Dynamic Mistral Tasks

### ISSUE THREAD

**Issue by nmaludy** [StackStorm/st2#3798]
## High-level

I'm trying to create a Mistral workflow that has `action` and `input` values specified dynamically via input parameters from the execution at runtime.

## Example Action Metadata

```
---
description: "Finds the host (VM) entry in Satellite"
enabled: true
entry_point: workflows/satellite_object_find.yaml
name: satellite_object_find
pack: encore
parameters:
  action_name:
    type: string
    description: "Name of the action use for finding the object with the following format 'foreman.<collection>.index' where <collection> is a foreman collection like hosts, hostgroups, organizations, etc."
    required: true
  extra_params:
    type: object
    description: "Additional parameters to add to be merged into the find request payload"
    required: false
  password:
    type: string
    description: "Password for login to Satellite"
    required: true
    secret: true
  server:
    type: string
    description: "Hostname/ip of the Satellite server."
    required: true
  search:
    type: string
    description: "Foreman API search query"
    required: true
  username:
    type: string
    description: "Username for login to Satellite"
    required: true
runner_type: "mistral-v2"
```

## Example Mistral Workflow
```
version: '2.0'

encore.satellite_object_find:
  description: >
    Finds the an object in Satellite, fail if no results are returned
  type: direct
  input:
    - action_name
    - extra_params
    - password
    - search
    - server
    - username
    - default_params:
        server: "{{ _.server }}"
        username: "{{ _.username }}"
        password: "{{ _.password }}"
        search: "{{ _.search }}"  
      
  output:
    obj: "{{ _.obj }}"
    id: "{{ _.id }}"
    
  tasks:
    init:
      action: std.noop
      publish:
        # merge the default_params and extra_params dicts
        obj_find_inputs: "{{ _.default_params.update(_.extra_params) if _.extra_params else ''}}{{ _.default_params }}"
      on-success:
        - obj_find
    
    obj_find:
      action: "{{ _.action_name }}"
      input: "{{ _.obj_find_inputs }}"
      on-success:
        - publish_vars: "{{ task('obj_find').result.result.results | length > 0 }}"
        - fail: "{{ task('obj_find').result.result.results | length == 0 }}"

    publish_vars:
      action: std.noop
      publish:
        obj: "{{ task('obj_find').result.result.results[0] }}"
        id: "{{ task('obj_find').result.result.results[0].id }}"
```

## What i'm trying to accomplish

In the [Foreman pack](https://github.com/StackStorm-Exchange/stackstorm-foreman) there are 199 different "collections" and each one has a different "index" action. These actions are used to find objects within the collection. 

The problem is that when the object is not found the server doesn't return an error, instead it simply returns an empty `results` array. This means that if we want to actually handle errors when objects aren't found we need to write logic to check for this condition and fail if it occurs.

After copy and pasting the "error handling" logic several times it got old so i decided to move it into an action. The idea was to create a "generic" action with the error handling logic that can be given the name of a Foreman index action to execute.

## Problem

When i try to run the workflow i get the following error:

```
error: Invalid action/workflow task property: {{ _.action_name }}
traceback:   File "/opt/stackstorm/st2/lib/python2.7/site-packages/st2actions/container/base.py", line 119, in _do_run
    (status, result, context) = runner.run(action_params)
  File "/opt/stackstorm/st2/lib/python2.7/site-packages/retrying.py", line 49, in wrapped_f
    return Retrying(*dargs, **dkw).call(f, *args, **kw)
  File "/opt/stackstorm/st2/lib/python2.7/site-packages/retrying.py", line 206, in call
    return attempt.get(self._wrap_exception)
  File "/opt/stackstorm/st2/lib/python2.7/site-packages/retrying.py", line 247, in get
    six.reraise(self.value[0], self.value[1], self.value[2])
  File "/opt/stackstorm/st2/lib/python2.7/site-packages/retrying.py", line 200, in call
    attempt = Attempt(fn(*args, **kwargs), attempt_number, False)
  File "/opt/stackstorm/runners/mistral_v2/mistral_v2.py", line 233, in run
    result = self.start_workflow(action_parameters=action_parameters)
  File "/opt/stackstorm/runners/mistral_v2/mistral_v2.py", line 258, in start_workflow
    def_dict_xformed = utils.transform_definition(def_dict)
  File "/opt/stackstorm/st2/lib/python2.7/site-packages/st2common/util/workflow/mistral.py", line 260, in transform_definition
    _transform_action(task_name, task_spec)
  File "/opt/stackstorm/st2/lib/python2.7/site-packages/st2common/util/workflow/mistral.py", line 200, in _transform_action
    _eval_inline_params(spec, action_key, input_key)
  File "/opt/stackstorm/st2/lib/python2.7/site-packages/st2common/util/workflow/mistral.py", line 118, in _eval_inline_params
    command, inputs = _parse_cmd_and_input(action_str)
  File "/opt/stackstorm/st2/lib/python2.7/site-packages/st2common/util/workflow/mistral.py", line 75, in _parse_cmd_and_input
    raise ValueError("Invalid action/workflow task property: %s" % cmd_str)
```

**Comment by nmaludy** [comment]
### Testing

FYI, this is perfectly valid in Mistral. I have the following example workflow:

``` yaml
version: '2.0'

test.workflow:
    description: A basic workflow that runs an arbitrary linux command.
    type: direct
    input:
        - action
        - output
          
    output:
        out: "{{ _.out }}"
        
    tasks:          
        task1:
            action: "{{ _.action }}"
            input:
                output: "{{ _.output }}"
            publish:
                out: "{{ task('task1').result }}"
```

And running the following commands:

``` shell
[root@stackstorm ~]# mistral workflow-create test_workflow.yaml 
+--------------------------------------+---------------+-----------+-------------------+--------+----------------+---------------------+------------+
| ID                                   | Name          | Namespace | Project ID        | Tags   | Input          | Created at          | Updated at |
+--------------------------------------+---------------+-----------+-------------------+--------+----------------+---------------------+------------+
| fdd9716b-5053-4834-bf5e-92162dae9eb8 | test.workflow |           | <default-project> | <none> | action, output | 2017-10-26 19:49:43 | None       |
+--------------------------------------+---------------+-----------+-------------------+--------+----------------+---------------------+------------+
[root@stackstorm ~]# mistral execution-create test.workflow '{"action": "std.echo", "output": "xxx123"}'
+--------------------+--------------------------------------+
| Field              | Value                                |
+--------------------+--------------------------------------+
| ID                 | e4152bb0-fca7-4bfc-8148-86d61b3b7aad |
| Workflow ID        | fdd9716b-5053-4834-bf5e-92162dae9eb8 |
| Workflow name      | test.workflow                        |
| Workflow namespace |                                      |
| Description        |                                      |
| Task Execution ID  | <none>                               |
| State              | RUNNING                              |
| State info         | None                                 |
| Created at         | 2017-10-26 19:49:52                  |
| Updated at         | 2017-10-26 19:49:52                  |
+--------------------+--------------------------------------+
[root@stackstorm ~]# mistral execution-get-output e4152bb0-fca7-4bfc-8148-86d61b3b7aad
{
    "out": "xxx123"
}
```

**Comment by guptasakshi01** [comment]
Hi,
How can we accomplish the same in YAQL ?

**Comment by nmaludy** [comment]
As a workaround for this we have come up with the "dispatch" pattern:

```yaml
input:
  - action_name

tasks:
  dispatch:
    action: std.noop
    on-success:
      - task_a: "{{ _.action_name == 'pack.action_a' }}"
      - task_b: "{{ _.action_name == 'pack.action_b' }}"
      - fail: "{{ _.action_name not in ['pack.action_a', 'pack.action_b'] }}"

  task_a:
    action: pack.action_a
  
  task_b:
    action: pack.action_b
```

**Comment by virenderdubey** [comment]
I also have the same requirement and getting the same error. Does anyone find the solution except that matching each action in on-success parameters. 
```
Invalid action/workflow task property: <% $.action %>
raise ValueError("Invalid action/workflow task property: %s" % cmd_str)
```

**Comment by m4dcoder** [comment]
This feature will be included in StackStorm new workflow engine, beta due v2.8. This feature will not be available before v2.8 and we will not be pursuing this feature in Mistral.
